src/presentation.ts

```diff
@@ -14,178 +22,223 @@ import {
-  ): PresentationRequest<Output, Inputs> {
-    const { presentationCircuitVKHash, action, serverNonce } = inputContext;
-    const claimsType = NestedProvable.fromValue(claims);
-    const claimsFields = Struct(claimsType).toFields(claims);
-    const claimsHash = Poseidon.hash(claimsFields);
+    context: { action: Field }
+  ) {
+    // generate random nonce on "the server"
+    let serverNonce = Field.random();
 
-    return {
-      type: 'zk-app',
+    return ZkAppRequest({
       spec,
       claims,
-      inputContext,
-      deriveContext: (walletContext: ZkAppWalletContext) => {
-        const { verifierIdentity, clientNonce } = walletContext;
-        const context = computeContext({
-          type: 'zk-app',
-          presentationCircuitVKHash,
-          clientNonce,
-          serverNonce,
-          verifierIdentity,
-          action,
-          claims: claimsHash,
-        });
-        return generateContext(context);
-      },
-    };
+      program: createProgram(spec),
+      inputContext: { ...context, type: 'zk-app', serverNonce },
+    });
   },
 
-  https<Output, Inputs extends Record<string, Input>>(
+  noContext<Output, Inputs extends Record<string, Input>>(
     spec: Spec<Output, Inputs>,
-    claims: Claims<Inputs>,
-    inputContext: HttpsInputContext
-  ): PresentationRequest<Output, Inputs> {
-    const { presentationCircuitVKHash, action, serverNonce } = inputContext;
-    const claimsType = NestedProvable.fromValue(claims);
-    const claimsFields = Struct(claimsType).toFields(claims);
-    const claimsHash = Poseidon.hash(claimsFields);
-
+    claims: Claims<Inputs>
+  ): NoContextRequest<Output, Inputs> {
     return {
-      type: 'https',
+      type: 'no-context',
       spec,
       claims,
-      inputContext,
-      deriveContext: (walletContext: HttpsWalletContext) => {
-        const { verifierIdentity, clientNonce } = walletContext;
-        const context = computeContext({
-          type: 'https',
-          presentationCircuitVKHash,
-          clientNonce,
-          serverNonce,
-          verifierIdentity,
-          action,
-          claims: claimsHash,
-        });
-        return generateContext(context);
-      },
+      inputContext: undefined,
+      deriveContext: () => Field(0),
     };
   },
+
   toJSON(request: PresentationRequest) {
-    return JSON.stringify(serializePresentationRequest(request));
+    let json = {
+      type: request.type,
+      spec: convertSpecToSerializable(request.spec),
+      claims: serializeNestedProvableValue(request.claims),
+      inputContext: serializeInputContext(request.inputContext),
+    };
+    return JSON.stringify(json);
   },
-  fromJSON<P extends PresentationRequest = PresentationRequest>(
-    json: string
-  ): P {
-    return deserializePresentationRequest(JSON.parse(json)) as P;
+
+  fromJSON<
+    R extends RequestFromType<K>,
+    K extends PresentationRequestType = PresentationRequestType
+  >(expectedType: K, json: string): R {
```

Comment:
when parsing a request from JSON, you have to specify the type of request that you accept. that is good for security and also lets us derive the correct type (using `RequestFromType`) for the wallet context